File: Python/matrix-diagonal-sum.py
# Time:  O(n)
# Space: O(1)

class Solution(object):
    def diagonalSum(self, mat):
        """
        :type mat: List[List[int]]
        :rtype: int
        """
        return sum(mat[i][i]+mat[~i][i] for i in xrange(len(mat))) - (mat[len(mat)//2][len(mat)//2] if len(mat)%2 == 1 else 0)

File: C++/matrix-diagonal-sum.cpp
// Time:  O(n)
// Space: O(1)

class Solution {
public:
    int diagonalSum(vector<vector<int>>& mat) {
        const int n = mat.size();
        int result = 0;
        for (int i = 0; i < n; ++i) {
            result += mat[i][i] + mat[n - 1 - i][i];
        }
        if (n % 2) {
            result -= mat[n / 2][n / 2];
        }
        return result;
    }
};
